apps/utils/prompt.py

```diff
@@ -14,10 +32,14 @@ def validate_prompt_variables(form_data, prompt_key: str, known_vars: set):
 
     if not form_data.get("source_material") and "source_material" in prompt_variables:
         raise ValidationError({prompt_key: "Prompt expects source_material but it is not provided."})
+    elif form_data.get("source_material") and "source_material" not in prompt_variables:
+        raise ValidationError({prompt_key: "source_material variable expected since source material is specified"})
+
+    if tools := form_data.get("tools", []):
+        tools_need = []
+        [tools_need.extend(PROMPT_VARS_REQUIRED_BY_TOOL[AgentTools(tool_name)]) for tool_name in tools]
```

Comment:
It's not great practice to cause side effects in list comprehension. Either make this a for loop or use `itertools.chain`:

```python
tools_need = list(chain.from_iterable(PROMPT_VARS_REQUIRED_BY_TOOL[AgentTools(tool_name)] for tool_name in tools))
```

Reply:
https://github.com/dimagi/open-chat-studio/pull/802/commits/3238fb0d589245d5374c8552a50d13871a0000cc